extensions/google-cloud-functions/runtime/src/main/java/io/quarkus/gcp/functions/QuarkusBackgroundFunction.java

```diff
@@ -0,0 +1,104 @@
+                String[] args = {};
+                Application app = (Application) appClass.getConstructor().newInstance();
+                app.start(args);
+                errorWriter.println("Quarkus bootstrapped successfully.");
+                started = true;
+            } catch (Exception ex) {
+                errorWriter.println("Quarkus bootstrap failed.");
+                ex.printStackTrace(errorWriter);
+            } finally {
+                Thread.currentThread().setContextClassLoader(currentCl);
+            }
+        } else {
+            errorWriter.println("Quarkus bootstrapped successfully.");
+            started = true;
+        }
+        deploymentStatus = error.toString();
+    }
+
+    static void setDelegates(String selectedDelegate, String selectedRawDelegate) {
+        if (selectedDelegate != null) {
+            try {
+                Class<?> clazz = Class.forName(selectedDelegate);
+                for (Method method : clazz.getDeclaredMethods()) {
+                    if (method.getName().equals("accept")) {
+                        // the first parameter of the accept method is the event, we need to register it's type to
+                        // be able to deserialize to it to mimic what a BackgroundFunction does
+                        if (method.getParameterTypes()[0] != Object.class) {// FIXME we have two accept methods !!!
+                            parameterType = method.getParameterTypes()[0];
+                        }
+                    }
+                }
+                delegate = (BackgroundFunction) Arc.container().instance(clazz).get();
+            } catch (ClassNotFoundException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        if (selectedRawDelegate != null) {
+            try {
+                Class<?> clazz = Class.forName(selectedRawDelegate);
+                rawDelegate = (RawBackgroundFunction) Arc.container().instance(clazz).get();
+            } catch (ClassNotFoundException e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    @Override
+    public void accept(String event, Context context) throws Exception {
+        if (!started) {
+            throw new IOException(deploymentStatus);
+        }
+
+        // TODO maybe we can check this at static init
+        if ((delegate == null && rawDelegate == null) || (delegate != null && rawDelegate != null)) {
+            throw new IOException("We didn't found any BackgroundFunction or RawBackgroundFunction to run " +
+                    "(or there is multiple one and none selected inside your application.properties)");
+        }
+
+        if (rawDelegate != null) {
+            rawDelegate.accept(event, context);
+        } else {
+            // TODO maybe use Gson as it's the library used by GCF
+            ObjectMapper mapper = new ObjectMapper();
+            Object eventObj = mapper.readValue(event, parameterType);
```

Comment:
This could happen every request?  Creation of the reader is an expensive operation and should be moved to static initialization within a @Recorder.  This will become more important if GCF ends up supporting a native binding.

Reply:
In fact I'm asking myself if using Jackson is really the way to implement it. GCF uses Gson for this maybe I need also to use it.

What is a little tricky here is that instead of implementing the three function types (HttpFunction, BackgroundFunctin, RawBakgroundFunction), I only implements two by Quarkus (HttpFunction and RawBackgroundFunction) as BackgroundFunction is the same as RawBakgroundFunction with the input event deserialized to the type parameter. So I choose to limit code duplication to deserialize this type parameter myself instead of letting GCP doing it for me and providing only one implementation.

Reply:
Is Funqy layered on top of this?  One thing to consider is that Funqy needs to be consistent across all cloud platform on what JSON marshalling layer it uses.  We can't use GSON for Funqy GCF and Jackson for Funqy Lambda.

Reply:
No, Funky uses it's own invoker.
I plan to refactor the three GCF functions extensions to share some code but not the marshalling layer.

Reply:
@patriot1burke I choose to switch to Gson as it is what GCF does inside it's invoker so let's be consistent.